src/charm.py

```diff
@@ -105,228 +110,145 @@ def _on_pgbouncer_pebble_ready(self, event: PebbleReadyEvent) -> None:
-
-        It should be made obvious to the user that this does not restart the container nor the
-        charm - it *only* restarts the application itself.
-        """
-        self._reload_pgbouncer()
-        event.set_results({"result": "pgbouncer application has restarted"})
-
     # ===================================
-    #  User management support functions
+    #  PgBouncer-Specific Utilities
     # ===================================
 
-    def _push_container_config(self, users: Dict[str, str] = None) -> None:
-        """Updates config files stored on pgbouncer container and reloads application.
-
-        Updates userlist.txt and pgbouncer.ini config files, reloading pgbouncer application once
-        updated.
-
-        Args:
-            users: a dictionary of usernames and passwords
-        """
-        if users is None:
-            users = self._get_userlist_from_container()
-        else:
-            self._push_userlist(users)
-
-        self._push_pgbouncer_ini(users, reload_pgbouncer=True)
-
-    def _push_pgbouncer_ini(self, users: Dict[str, str] = None, reload_pgbouncer=False) -> None:
+    def _render_pgb_config(self, config: pgb.PgbConfig, reload_pgbouncer=False) -> None:
         """Generate pgbouncer.ini from juju config and deploy it to the container.
 
         Args:
-            users: a dictionary of usernames and passwords
+            config: PgbConfig object containing pgbouncer config.
             reload_pgbouncer: A boolean defining whether or not to reload the pgbouncer application
                 in the container. When config files are updated, pgbouncer must be restarted for
                 the changes to take effect. However, these config updates can be done in batches,
                 minimising the amount of necessary restarts.
         """
-        if users is None:
-            users = self._get_userlist_from_container()
-
-        pgb_container = self.unit.get_container(self._pgbouncer_container)
-        pgbouncer_ini = self._generate_pgbouncer_ini(users)
-
-        try:
-            # Check that we're not updating this file unnecessarily
-            if pgb_container.pull(INI_PATH).read() == pgbouncer_ini:
-                logger.info("updated config does not modify existing pgbouncer config")
-                return
-        except FileNotFoundError:
-            # There is no existing pgbouncer.ini file, so carry on and add one.
-            pass
+        pgb_container = self.unit.get_container(PGB)
+        if not pgb_container.can_connect():
+            self.unit.status = WaitingStatus(
+                "Unable to push config to container - container unavailable."
+            )
+            return
 
         pgb_container.push(
             INI_PATH,
-            pgbouncer_ini,
-            user=self._pgbouncer_user,
-            permissions=0o600,
+            config.render(),
+            user=PG_USER,
+            permissions=0o777,
```

Comment:
Is 777 really needed? PgBouncer by default creates this file with 600 when installed through apt.

Reply:
You're right, this has been reverted back to 600